### PR TITLE
Support routing for directed device graphs

### LINKED ISSUE #5863: [Routing] Add support for directed device graphs.
**Summarize the task**
Currently, if a user wants to route their circuit on a device that has asymmetric coupling (i.e. you can execute a CNOT(q1, q2) but not CNOT(q2, q1)) then the transformer [raises an error](https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/transformers/routing/route_circuit_cqc.py#L105). 
As an improvement, we can support directed device graphs using the swap decomposition: SWAP(0,1) <=>
CNOT(0,1), H(0), H(1), CNOT(0,1), H(0), H(1), CNOT(0,1)

Related issues: #5838

### PR DESCRIPTION
**Pull request by Vivek1106-04**
Enhance the `RouteCQC` transformer to allow directed device graphs.
The routing algorithm is the same as for bidirectional graphs,
except for unidirectional edges q1 -> q2 where inserted SWAP
operations are replaced by their Hadamard decomposition

    CNOT(q1, q2)
    H(q1), H(q2)
    CNOT(q1, q2)
    H(q1), H(q2)
    CNOT(q1, q2)

Fixes #5863

### CODE REVIEW

**Comment by Vivek1106-04** [comment]
Added additional tests to cover edge cases for directed device graphs and the **tag_inserted_swaps** parameter.

**Comment by Vivek1106-04** [comment]
Used pylint disable for argument-count warnings

Refactoring to reduce args would change public API, breaking
notebooks that test against released Cirq. Disables are the
pragmatic choice for API stability.

**Comment by Vivek1106-04** [comment]
@tanujkhattar ! can you review this pr.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 1]
```diff
@@ -357,12 +401,16 @@ def process_executable_two_qubit_ops(timestep: int) -> int:
...
 
                 for swap in chosen_swaps:
-                    inserted_swap = mm.mapped_op(
-                        ops.SWAP(mm.int_to_logical_qid[swap[0]], mm.int_to_logical_qid[swap[1]])
+                    # Use the new emit_swap method to handle directed edges
+                    cls.emit_swap(
+                        routed_ops[timestep],
+                        (
+                            mm.int_to_physical_qid[mm.logical_to_physical[swap[0]]],
+                            mm.int_to_physical_qid[mm.logical_to_physical[swap[1]]],
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: Can we still use `mm.int_to_logical_qid` like we did earlier?

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 2]
```diff
@@ -285,13 +282,60 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
         return two_qubit_ops, single_qubit_ops
 
+    @classmethod
+    def emit_swap(
+        cls,
+        circuit_ops: list[cirq.Operation],
+        qubit_pair: tuple[cirq.Qid, cirq.Qid],
+        device_graph: nx.Graph,
+        tag_inserted_swaps: bool = False,
+    ) -> None:
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: I think it'll be cleaner if you make the change "Replace a SWAP with a Directional SWAP" in the `route_circuit` method between Step-4 and Step-5. You can inspect whether a gate has a `RoutingTag` or not and replace it with a directional swap gate in the `routing_ops`. You can define a `map_func` and use `cirq.map_operations` in order to do this replacement. This will help make the API cleaner and communicate the fact that routing logic doesn't care about direction, and we just use this little trick to decompose the swap gate using directional gates in the end. 

AFAIR, the router tries to preserve moment structure so you can also choose to encapsulate the decomposition of directional swap in a `cirq.CircuitOperation` or define a new `DirectionalSwap` gate that performs the decomposition and use it here. But I don't have strong opinions here.

Also add a little comment to the class that explains how the directional graphs are handled (i.e. the routing logic is equivalent to that of an undirected graph; you just use a directional swap decomposition to decompose the swap gates into CNOTs and single qubit gates)

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py (hunk 1 above)
Comment (Vivek1106-04) [reply cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: Yes, we still use mm.int_to_logical_qid and mm.int_to_physical_qid exactly as before. The 
_route method uses mm to map logical qubits to physical qubits when emitting SWAPs, and the return statement in 
route_circuit builds the swap permutation map using mm.int_to_logical_qid[k]. The only change is that we now emit standard SWAP gates during routing, and the directional decomposition happens in a separate post-processing step using map_operations_and_unroll

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 3]
```diff
@@ -357,12 +433,16 @@ def process_executable_two_qubit_ops(timestep: int) -> int:
...
+                    swap_qubits = (
+                        mm.int_to_physical_qid[mm.logical_to_physical[swap[0]]],
+                        mm.int_to_physical_qid[mm.logical_to_physical[swap[1]]],
                     )
+                    swap_op = ops.SWAP(*swap_qubits)
                     if tag_inserted_swaps:
-                        inserted_swap = inserted_swap.with_tags(ops.RoutingSwapTag())
-                    routed_ops[timestep].append(inserted_swap)
+                        swap_op = swap_op.with_tags(ops.RoutingSwapTag())
+                    routed_ops[timestep].append(swap_op)
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: I believe this can be reverted to it's original form? Do we need this change?

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py (hunk 1 above)
Comment (tanujkhattar) [reply cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: Out of curiosity -- Was this is an AI generated answer? It feels like one (nothing wrong with it, I'm just curious :))

Comment (Vivek1106-04) [reply cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: when you specially asked about `mm.int_to_logical_qid` i thought did i break the code some where else. to cross check myself i used ai to understand where `mm` is used .

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py (hunk 3 above)
Comment (Vivek1106-04) [reply cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: will revert back - just felt like it was easier on the eyes.

**Comment by Vivek1106-04** [comment]
@pavoljuhas ! review .

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 4]
```diff
@@ -81,6 +81,15 @@ class RouteCQC:
                 the swap that minimises the cost and use it to update our logical to physical
                 mapping. Repeat from 3.1.
 
+    **Handling Directed Graphs:**
+        When the device_graph is directed (i.e., edges represent unidirectional CNOT constraints),
+        the routing logic still operates as if the graph were undirected. This is because SWAP
+        gates are logically symmetric regardless of underlying gate direction constraints.
+        After routing completes, any inserted SWAP gates tagged with `RoutingSwapTag` are
+        decomposed into a directional-aware sequence using the Hadamard trick:
+        ``SWAP = CNOT(ctrl,tgt) - H⊗H - CNOT(ctrl,tgt) - H⊗H - CNOT(ctrl,tgt)``
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: nit - tweak the section for formatting consistency and remove the possibly confusing detail on tagging:

```suggestion
    Handling Directed Graphs:

        When the device_graph is directed (e.g., edges represent unidirectional CNOT constraints),
        the routing logic still operates as if the graph were undirected. This is because SWAP
        gates are logically symmetric regardless of underlying gate direction constraints.
        After routing completes, any inserted SWAP gates are decomposed into a directional-aware
        sequence using the Hadamard trick:
        ``SWAP = CNOT(ctrl, tgt) - H⊗H - CNOT(ctrl, tgt) - H⊗H - CNOT(ctrl, tgt)``
```

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 5]
```diff
@@ -231,10 +236,17 @@ def route_circuit(
             tag_inserted_swaps=tag_inserted_swaps,
         )
 
-        # 5. Return the routed circuit by packing each inner list of ops as densely as possible and
-        # preserving outer moment structure. Also return initial map and swap permutation map.
+        # 4.5. Replace tagged SWAP gates with directional decompositions if needed.
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: nit - renumber to 5 and below to 6, etc.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 6]
```diff
@@ -231,10 +236,17 @@ def route_circuit(
...
+        # This handles directed device graphs by decomposing SWAP into a sequence of CNOTs
+        # that respect the edge direction constraints.
+        routed_circuit = circuits.Circuit(circuits.Circuit(m) for m in routed_ops)
+        final_circuit = self._replace_swaps_with_directional_decomposition(
+            routed_circuit, tag_inserted_swaps
+        )
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: Please add a `if nx.is_directed(self.device_graph):` guard here so that we do not call possibly expensive `_replace_swaps_with_directional_decomposition` when it is a no-op.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 7]
```diff
@@ -285,6 +297,70 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
+            tag_inserted_swaps: Whether SWAPs were tagged during routing.
+
+        Returns:
+            Circuit with directional SWAP decompositions where needed.
+        """
+        if not tag_inserted_swaps:
+            # If swaps weren't tagged, we can't identify which ones to decompose.
+            # In this case, we assume the graph is undirected or the user will
+            # handle decomposition elsewhere.
+            return circuit
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: The resulting circuit should not depend on `tag_inserted_swaps` as that is a markup-only flag for whether the new operations should be tagged or not.

I suggest to construct a set of routing-added SWAP operations in the caller and pass it here.  The decomposition should happen for those SWAPs only.  The `tag_inserted_swaps` argument can be removed, because line 357 already copies the tags from the decomposed operation.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 8]
```diff
@@ -285,6 +297,70 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
+
+        def map_func(op: cirq.Operation, _: int) -> cirq.OP_TREE:
+            """Map function to replace tagged SWAPs with directional decomposition."""
+            # Check if this is a tagged SWAP operation
+            if not isinstance(op.gate, ops.SwapPowGate) or op.gate.exponent != 1:
+                return op
+            if not any(isinstance(tag, ops.RoutingSwapTag) for tag in op.tags):
+                return op
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: Please remove, the decomposition should happen also with `tag_inserted_swaps=False` when no tags are added.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 9]
```diff
@@ -285,6 +297,70 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
+
+            q1, q2 = op.qubits
+            has_forward = self.device_graph.has_edge(q1, q2)
+            has_reverse = self.device_graph.has_edge(q2, q1)
+
+            if has_forward and has_reverse:
+                # Bidirectional: keep the SWAP as-is
+                return op
+
+            if has_forward or has_reverse:
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py]: I suggest to avoid double handling of the bidirectional case and have only one fallback when it is not a directed edge:

```suggestion
            if has_forward ^ has_reverse:
```

File: cirq-core/cirq/transformers/routing/route_circuit_cqc.py [hunk 10]
```diff
@@ -285,6 +297,70 @@ def _get_one_and_two_qubit_ops_as_timesteps(
...
+                    ops.CNOT(ctrl, tgt),
+                ]
+                # Transfer tags from original SWAP to decomposed operations
+                return [op_i.with_tags(*op.tags) for op_i in decomposed_ops]
+
+            # Bidirectional or no edge check needed at routing level - keep as-is
+            return op
+
+        return transformer_primitives.map_operations_and_unroll(circuit, map_func)
+
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc.py:355]: On an unrelated note - please apply the patch below to fix pre-existing typos:
```patch
diff --git a/cirq-core/cirq/transformers/routing/route_circuit_cqc.py b/cirq-core/cirq/transformers/routing/route_circuit_cqc.py
index b19e0257b..0397d4d63 100644
--- a/cirq-core/cirq/transformers/routing/route_circuit_cqc.py
+++ b/cirq-core/cirq/transformers/routing/route_circuit_cqc.py
@@ -190 +190 @@ class RouteCQC:
-            tag_inserted_swaps: whether or not a RoutingSwapTag should be attched to inserted swap
+            tag_inserted_swaps: whether or not a RoutingSwapTag should be attached to inserted swap
@@ -375 +375 @@ class RouteCQC:
-        The i'th element of the returned list corresponds to the routed operatiosn in the i'th
+        The i'th element of the returned list corresponds to the routed operations in the i'th
@@ -385 +385 @@ class RouteCQC:
-        tag_inserted_swaps: whether or not a RoutingSwapTag should be attched to inserted swap
+        tag_inserted_swaps: whether or not a RoutingSwapTag should be attached to inserted swap

```

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 11]
```diff
@@ -22,8 +23,43 @@
...
     device = cirq.testing.construct_ring_device(10, directed=True)
     device_graph = device.metadata.nx_graph
-    with pytest.raises(ValueError, match="Device graph must be undirected."):
-        cirq.RouteCQC(device_graph)
+    # Directed graphs should now be accepted
+    router = cirq.RouteCQC(device_graph)
+    # Test that we can route a simple circuit on a directed graph
+    q = cirq.LineQubit.range(3)
+    circuit = cirq.Circuit(cirq.CNOT(q[0], q[1]), cirq.CNOT(q[1], q[2]))
+    hard_coded_mapper = cirq.HardCodedInitialMapper({q[i]: q[i] for i in range(3)})
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py]: nit - please correct here and below to be more clear that the mapper is identical
```suggestion
    hard_coded_mapper = cirq.HardCodedInitialMapper(dict(zip(q, q)))
```

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 12]
```diff
@@ -22,8 +23,43 @@
...
+    routed_circuit = router(circuit, initial_mapper=hard_coded_mapper)
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:32]: The CNOTs in the `circuit` satisfy edge directions so the `routed_circuit` is the same.  If intentional, please add `cirq.testing.assert_same_circuits(routed_circuit, circuit)`.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 13]
```diff
@@ -237,6 +273,148 @@ def test_empty_circuit() -> None:
...
+    hard_coded_mapper = cirq.HardCodedInitialMapper({q[i]: q[i] for i in range(4)})
+    # Need tag_inserted_swaps=True for directionality decomposition
+    routed_circuit, _, _ = router.route_circuit(
+        circuit, initial_mapper=hard_coded_mapper, tag_inserted_swaps=True
+    )
+
+    # Verify Hadamard gates are present
+    ops_list = list(routed_circuit.all_operations())
+    has_hadamards = any(op.gate == cirq.H for op in ops_list)
+    assert has_hadamards, "Expected Hadamard gates for reverse-only edge decomposition!"
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py]: nit - avoid obvious or repeated comments that are likely go stale with time

```suggestion
    # Verify Hadamard gates are present
    assert any(op.gate == cirq.H for op in routed_circuit.all_operations())
```

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 14]
```diff
@@ -237,6 +273,148 @@ def test_empty_circuit() -> None:
...
+
+
+def test_directed_device_with_tag_inserted_swaps_reverse_only() -> None:
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py]: This test is essentially the same as the `test_directed_device_reverse_only_edge` above.  Please combine.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 15]
```diff
@@ -237,6 +273,148 @@ def test_empty_circuit() -> None:
...
+
+    # Should have tagged Hadamards and CNOTs from the decomposed swap
+    assert len(tagged_ops) > 0, "No operations were tagged with RoutingSwapTag!"
+
+    # Verify both Hadamards and CNOTs are tagged
+    tagged_hadamards = [op for op in tagged_ops if op.gate == cirq.H]
+    tagged_cnots = [op for op in tagged_ops if isinstance(op.gate, cirq.CNotPowGate)]
+
+    assert len(tagged_hadamards) > 0, "Expected tagged Hadamard gates!"
+    assert len(tagged_cnots) > 0, "Expected tagged CNOT gates!"
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py]: As a general style thing - either use `assert tagged_hadamards` to check if non-empty or compare with the actual expected length.  Also please avoid non-essential assertion messages.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 16]
```diff
@@ -237,6 +273,148 @@ def test_empty_circuit() -> None:
...
+    # Circuit that needs a SWAP on the unidirectional edge
+    circuit = cirq.Circuit(cirq.CNOT(q[1], q[0]))
+
+    hard_coded_mapper = cirq.HardCodedInitialMapper({q[0]: q[0], q[1]: q[1]})
+    routed_circuit, initial_map, swap_map = router.route_circuit(
+        circuit, initial_mapper=hard_coded_mapper, tag_inserted_swaps=True
+    )
+
+    # Verify the routed circuit is mathematically equivalent to the original
+    cirq.testing.assert_circuits_have_same_unitary_given_final_permutation(
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:341]: I'd suggest to move this check to `test_directed_device_reverse_only_edge` which will give it a bit more check if that test routes a circuit with bidirectional and directed edges.

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 17]
```diff
@@ -237,6 +273,148 @@ def test_empty_circuit() -> None:
...
+    # Create a circuit with a tagged SWAP on q0-q2 (not in graph)
+    # This tests the fallback case in _replace_swaps_with_directional_decomposition
+    swap_on_non_edge = cirq.SWAP(q[0], q[2]).with_tags(cirq.RoutingSwapTag())
+    circuit = cirq.Circuit(swap_on_non_edge)
+
+    # Directly call the private method to test the fallback path
+    result = router._replace_swaps_with_directional_decomposition(circuit, True)
+
+    # The SWAP should be unchanged since there's no edge between q0 and q2
+    assert list(result.all_operations()) == [swap_on_non_edge]
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py]: This is testing impossible condition.  While you can call the internal function with any arguments, in the actual code the `_replace_swaps_with_directional_decomposition` is passed an already routed circuit which can have only q[0], q[1] from device_graph.  Comparing with operation that uses other qubit q[2] is not sensible.

Please either compare with expected routing result or if easier just delete this test.

**Review by pavoljuhas — CHANGES_REQUESTED** [review]
The routing result should not depend on the `tag_inserted_swaps` value.  That argument is a flag for tagging the new operations or not, but otherwise it should have no effect.

Also, please see the inline comments.

**Comment by Vivek1106-04** [comment]
done !,  removed some tests and added a new test for 100% coverage lines

File: cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py [hunk 18]
```diff
@@ -34,7 +34,8 @@ def test_directed_device() -> None:
     cirq.testing.assert_same_circuits(routed_circuit, circuit)
```
Comment (pavoljuhas) [inline cirq-core/cirq/transformers/routing/route_circuit_cqc_test.py:36]: Please parametrize this test for possible values of `tag_inserted_swaps`.  The expected circuit should be the same with the exception of tagging.  That can be handled as
```
t = [cirq.RoutingSwapTag()] if tag_inserted_swaps else []
# and then
cirq.H(q[0]).with_tags(*t)
....
```
